File: libwebrtc-sys/src/peer_connection_factory.cc
#include "iostream"
#include "rust/cxx.h"
#include "libwebrtc-sys/src/shared_bridge.rs.h"
#include "libwebrtc-sys/src/peer_connection_factory.rs.h"
#include "libwebrtc-sys/include/peer_connection_factory.h"
#include "libwebrtc-sys/include/api_internal.h"
#include "libwebrtc-sys/include/peer_connection_observer.h"

ArcasPeerConnectionFactory::ArcasPeerConnectionFactory(
    rtc::scoped_refptr<ArcasAPIInternal> internal_api,
    rtc::scoped_refptr<webrtc::PeerConnectionFactoryInterface> api) : internal_api(internal_api), api(api)
{
    internal_api->AddRef();
};

std::unique_ptr<webrtc::PeerConnectionInterface::RTCConfiguration> create_rtc_configuration(ArcasPeerConnectionConfig config)
{
    auto rtc = std::make_unique<webrtc::PeerConnectionInterface::RTCConfiguration>();
    webrtc::PeerConnectionInterface::IceServers servers;

    rtc->sdp_semantics = config.sdp_semantics;
    rtc->servers = servers;

    for (auto server_config : config.ice_servers)
    {
        webrtc::PeerConnectionInterface::IceServer rtc_ice_server;
        std::vector<std::string> rtc_urls;

        for (auto url : server_config.urls)
        {
            auto rtc_url = std::string(url.c_str());
            rtc_urls.push_back(rtc_url);
        }

        rtc_ice_server.urls = rtc_urls;
        rtc_ice_server.username = std::string(server_config.username.c_str());
        rtc_ice_server.password = std::string(server_config.password.c_str());
        servers.push_back(rtc_ice_server);
    }
    RTC_LOG(LS_VERBOSE) << "RTC LOG WITH " << servers.size() << " URLS";

    return rtc;
}

File: libwebrtc-sys/src/peer_connection_factory.rs
#[cxx::bridge]
pub mod ffi {
    #[derive(Debug, Clone)]
    struct ArcasICEServer {
        urls: Vec<String>,
        username: String,
        password: String,
    }

    #[derive(Debug)]
    struct ArcasPeerConnectionConfig {
        ice_servers: Vec<ArcasICEServer>,
        sdp_semantics: ArcasSDPSemantics,
    }

    unsafe extern "C++" {
        include!("include/alias.h");
        include!("include/peer_connection_factory.h");
        type ArcasPeerConnectionFactory;
        type ArcasRTCConfiguration = crate::shared_bridge::ffi::ArcasRTCConfiguration;
        type ArcasPeerConnectionObserver =
            crate::peer_connection_observer::ffi::ArcasPeerConnectionObserver;
        type ArcasPeerConnection = crate::peer_connection::ffi::ArcasPeerConnection;
        type ArcasVideoTrackSource = crate::video_track_source::ffi::ArcasVideoTrackSource;
        type ArcasVideoTrack = crate::video_track::ffi::ArcasVideoTrack;
        type ArcasSDPSemantics = crate::shared_bridge::ffi::ArcasSDPSemantics;

        fn create_rtc_configuration(
            config: ArcasPeerConnectionConfig,
        ) -> UniquePtr<ArcasRTCConfiguration>;

        fn gen_unique_peer_connection_factory() -> UniquePtr<ArcasPeerConnectionFactory>;

        // ArcasPeerConnectionFactory
        /// PeerConnection objects are threadsafe and can be shared between threads.
        /// the actual work happens on the worker thread.
        ///
        /// # Safety
        ///
        /// The observer must be kept alive as long as the peer connection object.
        ///
        unsafe fn create_peer_connection(
            self: &ArcasPeerConnectionFactory,
            config: UniquePtr<ArcasRTCConfiguration>,
            observer: *mut ArcasPeerConnectionObserver,
        ) -> SharedPtr<ArcasPeerConnection>;

        fn create_video_track(
            self: &ArcasPeerConnectionFactory,
            id: String,
            source: &ArcasVideoTrackSource,
        ) -> UniquePtr<ArcasVideoTrack>;
    }
}
